doc/admin-guide/layer-4-routing.en.rst

```diff
@@ -0,0 +1,124 @@
+matched against the configuration data to select an action for the inbound connection. In this case
+the option of interest is ``tunnel_route``. If this is set then |TS| will connect to the specified
+destination and issue an HTTP ``CONNECT`` request using the SNI value as the URL for the request.
+Because the destination and the ``CONNECT`` are the same in general it will be necessary to use
+a plugin to change the URL in the ``CONNECT``.
+
+Example
+-------
+
+Consider a Content Delivery Network (CDN) that has an edge layer of externally facing |TS|
+instances. The goal is to enable external clients to connect to internal services and do their own
+client certificate verification, possibly because distribution of private keys to the edge |TS|
+instances is too difficult or too risky. To achieve this, the edge |TS| instances can be configured
+to route inbound TLS connections with specific SNI values directly to the internal services without
+TLS termination on the edge. This enables the edge to provide controlled external access to the
+internal services without each internal service having to stand up its own edge. Note the services
+do not require global routable addresses as long as the edge |TS| instances can route to the
+services.
+
+The basic set up is therefore
+
+.. figure:: ../uml/images/l4-example-cdn-layout.svg
+   :align: center
+
+   A Client connects to an edge |TS| which forwards the connection to the internal Service.
+   The Client then negotiates TLS with the Service.
+
+
+For the example, let us define two services inside the corporate network of Example, Inc. ``service-1`` and
+``service-2``. ``service-1`` is on port 443 on host ``app-server-29`` while ``service-2`` is on port
+4443 on host ``app-server-56``. The SNI routing set up for this would be
+
+========================== =====================================
+SNI value                  Destination
+========================== =====================================
+service-1.example.com      app-server-29:443
+service-2.example.com      app-server-56:4443
+========================== =====================================
+
+The :file:`ssl_server_name.config` contents would be
+
+.. code-block:: lua
+
+   server_config = {
+      {
+         fqdn = 'service-1.example.com'
+         tunnel_route = 'app-server-29:443'
+      },
+      {
+         fqdn = 'service-2.example.com'
+         tunnel_route = 'app-server-56:4443'
+      }
+   }
+
+The sequence of network activity for a Client connecting to ``service-2`` is
+
+.. image:: ../uml/images/l4-sni-routing-seq.svg
+   :align: center
+
+Note the destination for the outbound TCP connection and the HTTP ``CONNECT`` is the same. If this
+is a problem (which it will be in general) a plugin is needed to change the URL in the ``CONNECT``.
+In this case the proxy request is available in the :c:macro:`TS_HTTP_TXN_START_HOOK` hook. This
+cannot be done using remap because for a ``CONNECT`` there is no remap phase.
```

Comment:
I think it is obvious, but still I feel like we need to mention that plugins hooked at post transaction start hooks  will not work. However, TXN_CLOSE_HOOK  gets triggered. Also there will be no caching

Reply:
Fixed.